0.12 TypeMap Validation Issue

It looks like TypeMap forces the configuration to specify `key` when instantiating a TypeMap blockin 0.12. 
```
    testing.go:568: Step 0 error: config is invalid: Missing key for permissions: All permissions blocks must have 1 labels (key).
``` 

There is a [resource](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_ssm_document.go#L129) in the aws provider that uses different names for the keys. Is this functionality being lost with the swap to 0.12?

> Would one of you mind doing a quick test to confirm whether my guess is correct here that the keys of these maps are not being validated at all?

```hcl
resource "aws_cognito_identity_pool_roles_attachment" "main" {
  identity_pool_id = "${aws_cognito_identity_pool.main.id}"

  roles {
    "authenticated" = "${aws_iam_role.authenticated.arn}"
    "invalid"       = "${aws_iam_role.authenticated.arn}"
  }
}
```

No schema validation error -- fails trying to send `invalid` over the wire.

```
--- FAIL: TestAccAWSCognitoIdentityPoolRolesAttachment_basic (8.44s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
        	* aws_cognito_identity_pool_roles_attachment.main: 1 error occurred:
        	* aws_cognito_identity_pool_roles_attachment.main: Error creating Cognito Identity Pool Roles Association: ValidationException: 1 validation error detected: Value '{authenticated=arn:aws:iam::187416307283:role/cognito_authenticated_6w1h21x0qi, invalid=arn:aws:iam::187416307283:role/cognito_authenticated_6w1h21x0qi}' at 'roles' failed to satisfy constraint: Map keys must satisfy constraint: [Member must satisfy regular expression pattern: (un)?authenticated]
```

--- a/helper/schema/core_schema.go
+++ b/helper/schema/core_schema.go
@@ -39,6 +39,21 @@ func (m schemaMap) CoreConfigSchema() *configschema.Block {
 			ret.Attributes[name] = schema.coreConfigSchemaAttribute()
 			continue
 		}
+		if schema.Type == TypeMap {
+			// For TypeMap in particular, it isn't valid for Elem to be a
+			// *Resource (since that would be ambiguous in flatmap) and
+			// so Elem is treated as a TypeString schema if so. This matches
+			// how the field readers treat this situation, for compatibility
+			// with configurations targeting Terraform 0.11 and earlier.
+			if _, isResource := schema.Elem.(*Resource); isResource {
+				sch := *schema // shallow copy
+				sch.Elem = &Schema{
+					Type: TypeString,
+				}
+				ret.Attributes[name] = sch.coreConfigSchemaAttribute()
+				continue
+			}
+		}
 		switch schema.Elem.(type) {
 		case *Schema, ValueType:
 			ret.Attributes[name] = schema.coreConfigSchemaAttribute()

--- a/helper/schema/core_schema_test.go
+++ b/helper/schema/core_schema_test.go
@@ -204,7 +204,18 @@ func TestSchemaMapCoreConfigSchema(t *testing.T) {
 				},
 			},
 			testResource(&configschema.Block{
-				Attributes: map[string]*configschema.Attribute{},
+				Attributes: map[string]*configschema.Attribute{
+					// This one becomes a string attribute because helper/schema
+					// doesn't actually support maps of resource. The given
+					// "Elem" is just ignored entirely here, which is important
+					// because that is also true of the helper/schema logic and
+					// existing providers rely on this being ignored for
+					// correct operation.
+					"map": {
+						Type:     cty.Map(cty.String),
+						Optional: true,
+					},
+				},
 				BlockTypes: map[string]*configschema.NestedBlock{
 					"list": {
 						Nesting:  configschema.NestingList,
@@ -216,10 +227,6 @@ func TestSchemaMapCoreConfigSchema(t *testing.T) {
 						Nesting:  configschema.NestingSet,
 						Block:    configschema.Block{},
 						MinItems: 1, // because schema is Required
-					},
-					"map": {
-						Nesting: configschema.NestingMap,
-						Block:   configschema.Block{},
 					},
 				},
 			}),

